### PR TITLE
INA260 Current and Power Sensor support

### PR DESCRIPTION
**Pull request by mreditor97**
# What does this implement/fix? 

Adding of INA260 Current and Power sensor support to ESP Home.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1649

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
i2c:
  sda: 21
  scl: 22
  scan: true

sensor:
  - platform: ina260
    address: 0x40
    current:
      name: "INA260 Current"
    power:
      name: "INA260 Power"
    bus_voltage:
      name: "INA260 Voltage"
    max_current: 15A
    update_interval: 60s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ina260/ina260.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+enum {
+  INA260_REGISTER_CONFIG = 0x00,
+  INA260_REGISTER_CURRENT = 0x01,
+  INA260_REGISTER_BUS_VOLTAGE = 0x02,
+  INA260_REGISTER_POWER = 0x03,
+  INA260_REGISTER_MASK_ENABLE = 0x06,
+  INA260_REGISTER_ALERT_LIMIT = 0x07,
+  INA260_REGISTER_MANUFACTURE_ID = 0xFE,
+  INA260_REGISTER_DEVICE_ID = 0xFF
+};
```
Comment (jesserockz) [inline esphome/components/ina260/ina260.cpp]: Can you move these to `static const uint8_t` please?

File: esphome/components/ina260/ina260.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+  // Reset device on setup
+  // 0bx000000000000000 << 15 RESET Bit (1 -> trigger reset)
+  if (!this->write_byte_16(INA260_REGISTER_CONFIG, 0x8000)) {
+    this->error_code_ = DEVICE_RESET_FAILED;
+    this->mark_failed();
+    return;
+  }
+
+  delay(2);  // NOLINT
```
Comment (jesserockz) [inline esphome/components/ina260/ina260.cpp]: Don't need to unnecessarily bypass lint here.

```suggestion
  delay(2);
```

File: esphome/components/ina260/ina260.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  uint16_t config = 0x0000;
+
+  // Averaging Mode AVG Bit Settings[11:9] (000 -> 1 sample, 001 -> 4 sample, 111 -> 1024 samples)
+  config |= 0b0000001000000000;
+
+  // Bus Voltage Conversion Time VBUSCT Bit Settings [8:6] (100 -> 1.1ms, 111 -> 8.244 ms)
+  config |= 0b0000000100000000;
+
+  // Mode Settings [2:0] Combinations (111 -> Shunt, Bus, Continuous)
+  config |= 0b0000000000000111;
```
Comment (jesserockz) [inline esphome/components/ina260/ina260.cpp]: If this is not configurable, you should just make it a const to save memory on the variable.

File: esphome/components/ina260/ina260.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  void set_max_current_amps(float max_current_amp) { this->max_current_amp_ = max_current_amp; }
+  void set_bus_voltage_sensor(sensor::Sensor *bus_voltage_sensor) { this->bus_voltage_sensor_ = bus_voltage_sensor; }
+  void set_current_sensor(sensor::Sensor *current_sensor) { this->current_sensor_ = current_sensor; }
+  void set_power_sensor(sensor::Sensor *power_sensor) { this->power_sensor_ = power_sensor; }
+
+ protected:
+  uint16_t manufacture_id_{0};
+  uint16_t device_id_{0};
+
+  float max_current_amp_{15};
```
Comment (jesserockz) [inline esphome/components/ina260/ina260.h]: Not used anywhere?

File: esphome/components/ina260/sensor.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            ),
+            cv.Optional(CONF_POWER): sensor.sensor_schema(
+                unit_of_measurement=UNIT_WATT,
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_POWER,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_MAX_CURRENT, default=15): cv.All(
+                cv.current, cv.Range(min=0.0)
+            ),
```
Comment (jesserockz) [inline esphome/components/ina260/sensor.py]: Not used?

File: esphome/components/ina260/ina260.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  if (this->current_sensor_ != nullptr) {
+    uint16_t raw_current;
+    if (!this->read_byte_16(INA260_REGISTER_CURRENT, &raw_current)) {
+      this->status_set_warning();
+      return;
+    }
+    float current_ma = int16_t(raw_current) * 0.00125f;
+    this->current_sensor_->publish_state(current_ma);
```
Comment (jesserockz) [inline esphome/components/ina260/ina260.cpp]: Is this mA or A? You have Amps in the python config and mA here...

File: esphome/components/ina260/ina260.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  if (this->power_sensor_ != nullptr) {
+    uint16_t raw_power;
+    if (!this->read_byte_16(INA260_REGISTER_POWER, &raw_power)) {
+      this->status_set_warning();
+      return;
+    }
+    float power_mw = ((int16_t(raw_power) * 10.0f) / 1000.0f);
+    this->power_sensor_->publish_state(power_mw);
```
Comment (jesserockz) [inline esphome/components/ina260/ina260.cpp]: Is this mW or W? You have Watts in the python config and mW here...

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/ina260/ina260.cpp (hunk 1 above)
Comment (mreditor97) [reply esphome/components/ina260/ina260.cpp]: I have changed them over to `static const uint8_t`.
However, can I ask your reasoning for this change? Any knowledge you have would be greatly appreciated. Thanks 👍

File: esphome/components/ina260/ina260.cpp (hunk 3 above)
Comment (mreditor97) [reply esphome/components/ina260/ina260.cpp]: Sorted this, merged all values into one and placed it directly where the config parameter was being used (It is only used once).

**Comment by mreditor97** [comment]
Accidentally closed the pull request, so managed to reopen it. Will update it again now it has failed the checks.

**Comment by mreditor97** [comment]
Is there something else that I need to change on this for it to be implemented?

File: esphome/components/ina260/ina260.cpp (hunk 1 above)
Comment (jesserockz) [reply esphome/components/ina260/ina260.cpp]: Consistency with the rest of the codebase =)

**Review by jesserockz — APPROVED** [review]
LGTM, thanks for making the changes and sorry it took so long to get back to it.

**Pull request merged**